src/Graph.ts

```diff
@@ -611,12 +616,35 @@ Try defining "${chunkName}" first in the manualChunks definitions of the Rollup
 		});
 	}
 
-	private fetchModule(id: string, importer: string): Promise<Module> {
+	private fetchModule(id: string, importer: string): Promise<Module | WasmModule> {
 		// short-circuit cycles
 		const existingModule = this.moduleById.get(id);
 		if (existingModule) {
 			if (existingModule.isExternal) throw new Error(`Cannot fetch external module ${id}`);
-			return Promise.resolve(<Module>existingModule);
+			return Promise.resolve(<Module | WasmModule>existingModule);
+		}
+
+		if (extname(id) === '.wasm') {
+			const module = new WasmModule(this, id);
+			this.moduleById.set(id, module);
+
+			// FIXME(sven): alternatively we can use a loadBinary method
+			return this.load(id)
+				.catch((err: Error) => {
+					// FIXME(sven): error handling here
+					throw err;
+				})
+				.then((data: string) => {
+					// TODO(sven): still check magic header here
+					const bin = new Buffer(data);
+					module.setSource(bin);
+
+					this.fetchAllDependencies(module);
+
+					this.modules.push(module);
+
+					return module;
+				});
```

Comment:
Don't worry about having this blatant hack here - let's leave this as a big TODO and come back to it later.

Reply:
Thinking about this a little more - instead of trying to introduce some new type of plugin or resolve API to distinguish WASM, let's just use the binary header bytes detection method for this. Saves a lot of pain!